ath79: lzma-loader: add option to boot alternate kernel image

mcprat opened the pull request:
Some boards that use the OKLI lzma-loader (like many Engenius models)
have a read-only failsafe image on flash that is loaded
when the bootloader detects that the kernel or rootfs is missing or damaged
via md5 checksums that are stored in the u-boot environment.

When OKLI is used in place of the kernel, and the real kernel
is missing or damaged, the system halts instead.
Because many of the boards' bootloaders have bugged and nonfunctional TFTP,
in this situation the only reliable way to recover is with UART serial console.

This adds logic to use an alternative flash offset value
defined with a makefile variable when the real kernel
with OKLI magic number cannot be found or fails to boot
and loads the alternative kernel from the flash instead of halting.

There are 3 steps for loading the kernel.
if any of them return with failure,
the alternative offset variable is defined to be the offset used
and the steps are attempted again without the for loop
that identifies the offset with the OKLI magic number.

The size of the lzma compressed image of OKLI lzma-loader
when not used for kernel-loader build command
has increased by    0x050  or    80 bytes
to a total size of  0xD51  or  3409 bytes
however the exact size depends on the makefile variables usage.

Usage example:
(where 0x670000 is the address of the failsafe image defined in DTS)

define Device/engenius_enh202-v1
  $(Device/engenius_loader_okli)
  SOC := ar7240
  DEVICE_MODEL := ENH202
  DEVICE_VARIANT := v1
  DEVICE_PACKAGES := rssileds
  IMAGE_SIZE := 4864k
  LOADER_FLASH_OFFS := 0x1b0000
  LOADER_FLASH_ALT := 0x670000
  ENGENIUS_IMGNAME := senao-enh202
endef

compiles successfully
tested on Engenius ENH202 v1

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
ath79: lzma-loader: add option to boot alternate image

Some boards that use the OKLI lzma-loader (like many Senao boards)
have a read-only failsafe image on flash that is loaded
when the bootloader detects that the kernel or rootfs is missing or damaged
via md5 checksums that are stored in the u-boot environment.

When OKLI is used in place of the kernel, and the real kernel
is missing or damaged, the system halts instead.
Because many of the boards' bootloaders have bugged and nonfunctional TFTP,
in this situation the only reliable way to recover is with UART serial console.

This adds logic to use an alternative flash offset value
defined with a makefile variable when the real kernel
with OKLI magic number cannot be found or fails to boot
and loads the alternative kernel from the flash instead of halting.

There are 3 steps for loading the kernel.
if any of them return with failure,
the alternative offset variable is defined to be the offset used
and the steps are attempted again without the for loop
that identifies the offset with the OKLI magic number
during the default attempt.

using #ifdef / #else / #endif in loader_main() to further save space

The size of the lzma compressed image of OKLI lzma-loader
when not used for kernel-loader build command
has increased by    0x050  or    80 bytes
to a total size of  0xD51  or  3409 bytes
however the exact size depends on the makefile variables usage.

Usage example:
(where 0x670000 is the address of the failsafe image defined in DTS)

define Device/engenius_enh202-v1
  $(Device/engenius_loader_okli)
  SOC := ar7240
  DEVICE_MODEL := ENH202
  DEVICE_VARIANT := v1
  DEVICE_PACKAGES := rssileds
  IMAGE_SIZE := 4864k
  LOADER_FLASH_OFFS := 0x1b0000
  LOADER_FLASH_ALT := 0x670000
  ENGENIUS_IMGNAME := senao-enh202
endef

compiles successfully
tested on Engenius ENH202 v1

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
I test by deleting the kernel and reboot

`mtd erase kernel`

before
```
Validating MD5Sum of 'vmlinux'... Passed!
Validating MD5Sum of 'rootfs'... Passed!
## Booting image at 9f0a0000 ...
   Image Name:   MIPS OpenWrt Linux-5.4.60
   Created:      2020-08-28  21:16:23 UTC
   Image Type:   MIPS Linux Kernel Image (lzma compressed)
   Data Size:    3265 Bytes =  3.2 kB
   Load Address: 80060000
   Entry Point:  80060000
   Verifying Checksum at 0x9f0a0040 ...OK
   Uncompressing Kernel Image ... OK
No initrd
## Transferring control to Linux (at address 80060000) ...
## Giving linux memsize in bytes, 33554432

Starting kernel ...


OpenWrt kernel loader for AR7XXX/AR9XXX
Copyright (C) 2011 Gabor Juhos <juhosg@openwrt.org>
Looking for OpenWrt image... not found!

System halted!
```

after
```
Validating MD5Sum of 'vmlinux'... Passed!
Validating MD5Sum of 'rootfs'... Passed!
## Booting image at 9f0a0000 ...
   Image Name:   MIPS OpenWrt Linux-5.4.72
   Created:      2020-10-29   1:51:57 UTC
   Image Type:   MIPS Linux Kernel Image (lzma compressed)
   Data Size:    3511 Bytes =  3.4 kB
   Load Address: 80060000
   Entry Point:  80060000
   Verifying Checksum at 0x9f0a0040 ...OK
   Uncompressing Kernel Image ... OK
No initrd
## Transferring control to Linux (at address 80060000) ...
## Giving linux memsize in bytes, 33554432

Starting kernel ...



OpenWrt kernel loader for AR7XXX/AR9XXX
Copyright (C) 2011 Gabor Juhos <juhosg@openwrt.org>
Looking for OpenWrt image... not found!
Booting alternate kernel image at 0xbf670000
Decompressing kernel... done!
Starting kernel at 80060000...

Booting AR7240(Python)...

```

mpratt14 commented:
I am still new to the C language so please give me a tough review. I don't know if I have to instantiate all the variables again in the function I made. If there is a more efficient way please advise me.

xabolcs commented:
> ...and allow FLASH_MAX to be set by a makefile variable for testing

I like the idea personally, but those helper changes should be kept in the developer's tree, IMHO.

xabolcs commented on target/linux/ath79/image/lzma-loader/src/loader.c:
```diff
@@ -228,7 +258,15 @@ void loader_main(unsigned long reg_a0, unsigned long reg_a1,
 	if (res != LZMA_RESULT_OK) {
 		printf("Incorrect LZMA stream properties!\n");
-		halt();
+		if (CONFIG_FLASH_ALT != 0) {
```
... and `LZMA_WRAPPER` is not defined ...

xabolcs commented on target/linux/ath79/image/lzma-loader/src/loader.c:
```diff
@@ -243,7 +281,22 @@ void loader_main(unsigned long reg_a0, unsigned long reg_a1,
 			printf("unknown error %d!\n", res);
 		}
-		halt();
+		if (CONFIG_FLASH_ALT != 0) {
```
... and `LZMA_WRAPPER` is not defined ...

xabolcs reviewed (commented):
This `lzma_init_data_alt()` addition could work for both loader type, but if you would like to hide it from `LZMA_WRAPPER` then hide it correctly! :wink:

mpratt14 commented:
Can someone explain how LZMA_WRAPPER is used if at all? Seems to never be defined in ath79...

I guess the assumption is that if LZMA_WRAPPER is used then FLASH_ALT is never set...

xabolcs commented:
Grep for `loader-kernel` for usage, and check its definition at `Build/loader-kernel`!
Follow the `LOADER_DATA`!

Its used for passing the decompression work from U-Boot to LZMA Loader.

mpratt14 commented:
Added #if/#else/#endif directives for LZMA_WRAPPER the same way it is for the function

xabolcs commented:
Now you made `lzma_init_data_alt()` hidden nicely, it would be nice to note in the commit title too, that this is an OKLI-only stuff.

While it could work for the `LZMA_WRAPPER` case, it makes more sense for the current case.

Another nitpicking from me is to include an example usage in the commit message, like what **981213** did recently: c063e27e02a9dcac0e7f5877fb154e58fa3e1a69, 9a168b51c3a60c95ccd349dd3010615ba6b0e033

mpratt14 commented:
Updated commit message

I think title is ok, I made it clear in commit message, but they can change the title at the end if they want

mpratt14 commented:
@NeoRaider If you have the time I would appreciate your review, thanks

mpratt14 commented:
@blogic would you be available to review?

mpratt14 commented:
@ynezz do you think it is possible that this can be accepted before new version is branched? or if you know anyone who is not too busy to review it?

mpratt14 commented:
@981213 would you be available to review?

blocktrron commented on target/linux/ath79/image/lzma-loader/src/loader.c:
```diff
@@ -243,7 +285,27 @@ void loader_main(unsigned long reg_a0, unsigned long reg_a1,
+#if (LZMA_WRAPPER)
 		halt();
+#else
+		if (CONFIG_FLASH_ALT != 0) {
```
We know at compile time whether or not there is an alternative location at all. No need to compile for single-image scenarios.

mpratt14 replied:
True, the only difference would be the size of the loader...you can decide whether that is important enough

blocktrron commented on target/linux/ath79/image/lzma-loader/src/loader.c:
```diff
@@ -243,7 +285,27 @@ void loader_main(unsigned long reg_a0, unsigned long reg_a1,
 		halt();
+#else
+		if (CONFIG_FLASH_ALT != 0) {
+			lzma_init_data_alt();
+			res = lzma_init_props();
+			if (res != LZMA_RESULT_OK) {
+				halt();
+			}
```
This seems redundant.

Also the code-flow looks not really great. If the first location is corrupt and the second one has an intact LZMA header but corrupt payload, you are effectively trying to boot twice from the same source.

Better: Abstract the boot process to a parameterized function which returns non-zero in case the LZMA decompression fails at some stage. If the loader is expected to try a second location from flash, it can reuse the entire boot process function.

mpratt14 replied:
thanks a lot for the review and ideas. I'll try some things to make this much cleaner looking

blocktrron reviewed (commented):
Well, personally I'd like to keep this lzma-loader hackery as slim as possible, given we have to maintain it for the target lifetime.

Having non-necessary special features for a small subset of devices in this stage is not something i think we should put ourselves up with, but someone might have a different opinion about that.

Anyways, here's what I've spotted.

mcprat committed:
ath79: lzma-loader: refactor code

In preparation for a patch to load an alternative kernel
if finding the openwrt kernel fails:

 - general organization
 - moving the 3 steps for loading the kernel
	into a separate function "kernel_load"
 - let halt() function only be called once, other cases return
 - let all return variables be called "ret"
 - removed some useless variables
 - replaced some conditional statements for clarity

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
ath79: implement LOADER_FLASH_ALT for Senao boards

These Senao boards have an extra "failsafe" image on flash
for recovery when the kernel fails to boot.

They also use the OKLI kernel loader, which originally
would halt when the kernel is missing or fails to boot.

Now, with this added feature and variable,
OKLI kernel loader will load the failsafe image instead of halting,
in case of the real kernel being damaged or missing,
which otherwise would require serial console for recovery.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
@blocktrron Ready for review again. The diff is messy but the result is much cleaner, and the size difference is very small now 👍🏽

mpratt14 commented:
switched to using #ifdef so that changes to config.h can be dropped

mpratt14 commented:
now implemented for 9 boards

I know one of the concerns is that it affects only a small set of devices
however, that number can grow up to possibly 25 or more as we get more Engenius products added over time...

and who knows it might be useful for something that's not an Engenius product

mpratt14 commented:
@svanheule if you have some time, is it possible you can test this on one of your eap2xx devices which uses `loader-kernel`?

the new feature does not affect them, but the code refactoring does. Also, ideas are welcome of course 👍🏼

svanheule commented:
I can at least confirm that these patches don't hurt `loader-kernel` on my EAP245v1; it still boots with these patches applied.

981213 commented:
Hi! I think it's clear that no maintainer really wants to further complicate current lzma-loader.
I'm closing this PR now, as it's been open here for 2 years and nobody decides to pick it up.
